### PR TITLE
Introduce pre redirect hook plugin during auth callback

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
# TL;DR
Introduces a preRedirectHook plugin handler to allow custom code to be able to be plugged in before the successful auth call back redirect happens

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

```
type PreRedirectHookFunc func(ctx context.Context, authCtx interfaces.AuthenticationContext, request *http.Request, w http.ResponseWriter) *PreRedirectHookError
```

PreRedirectHookFunc Interface is used for running custom code before the redirect happens during a successful auth flow.

The API allows you to access the authenticationContext, request and response 

This might be useful in cases where the auth flow allows the user to login since the IDP has been configured
for eg: to allow all users from a particular domain to login but you want to restrict access to only a particular set of user ids. eg : users@domain.com are allowed to login but user user1@domain.com, user2@domain.com should only be allowed

The current Plugin hook will allow you to add such a custom code.

PreRedirectHookError is the error interface which allows the user to set correct http status code and Message to be set in case the function returns an error without which the current usage in GetCallbackHandler will set this to InternalServerError

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/3940

## Follow-up issue
_NA_

### CODE REVIEW

File: auth/handlers.go [hunk 1]
```diff
@@ -178,7 +196,22 @@ func GetCallbackHandler(ctx context.Context, authCtx interfaces.AuthenticationCo
 			return
 		}
 
+		preRedirectHook := plugins.Get[PreRedirectHookFunc](pluginRegistry, plugins.PluginIDPreRedirectHook)
+		if preRedirectHook != nil {
+			logger.Infof(ctx, "preRedirect hook is set")
+			if err := preRedirectHook(ctx, authCtx, request, writer); err != nil {
+				logger.Errorf(ctx, "failed the preRedirect hook due %v with status code  %v", err.Message, err.Code)
+				if http.StatusText(err.Code) != "" {
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply auth/handlers.go:204]: So i am relying on the user code which set the hook to set the correct http status code .
If they choose to set OK, then thats what we will set on the header.

File: auth/handlers.go [hunk 2]
```diff
@@ -178,7 +196,22 @@ func GetCallbackHandler(ctx context.Context, authCtx interfaces.AuthenticationCo
...
+					writer.WriteHeader(err.Code)
+				} else {
+					writer.WriteHeader(http.StatusInternalServerError)
+				}
+				return
+			}
+			logger.Info(ctx, "Successfully called the preRedirect hook")
+		}
 		redirectURL := getAuthFlowEndRedirect(ctx, authCtx, request)
+		logger.Infof(ctx, "Going to perform the redirect with redirectURl %v", redirectURL)
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply auth/handlers.go:214]: Not needed. removed.

**Review by katrogan — APPROVED** [review]
